File: src/main/java/com/uiptv/server/api/json/HttpSeriesDetailsJsonServer.java
package com.uiptv.server.api.json;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.uiptv.db.SeriesEpisodeDb;
import com.uiptv.model.Account;
import com.uiptv.model.Channel;
import com.uiptv.service.AccountService;
import com.uiptv.service.ConfigurationService;
import com.uiptv.service.HandshakeService;
import com.uiptv.service.ImdbMetadataService;
import com.uiptv.shared.Episode;
import com.uiptv.shared.EpisodeList;
import com.uiptv.shared.SeasonInfo;
import com.uiptv.ui.XtremeParser;
import com.uiptv.util.AccountType;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.uiptv.util.ServerUtils.generateJsonResponse;
import static com.uiptv.util.ServerUtils.getParam;
import static com.uiptv.util.StringUtils.isBlank;

public class HttpSeriesDetailsJsonServer implements HttpHandler {
    @Override
    public void handle(HttpExchange ex) throws IOException {
        Account account = AccountService.getInstance().getById(getParam(ex, "accountId"));
        if (account == null) {
            generateJsonResponse(ex, "{\"seasonInfo\":{},\"episodes\":[]}");
            return;
        }
        if (account.isNotConnected()) {
            HandshakeService.getInstance().connect(account);
        }

        String seriesId = getParam(ex, "seriesId");
        String categoryId = getParam(ex, "categoryId");
        String seriesName = getParam(ex, "seriesName");
        JSONObject response = new JSONObject();
        response.put("seasonInfo", new JSONObject());
        response.put("episodes", new JSONArray());
        response.put("episodesMeta", new JSONArray());
        if (!isBlank(seriesId)) {
            List<Channel> cached = SeriesEpisodeDb.get().getEpisodes(account, categoryId, seriesId);
            if (!cached.isEmpty() && SeriesEpisodeDb.get().isFresh(account, categoryId, seriesId, ConfigurationService.getInstance().getCacheExpiryMs())) {
                response.put("episodes", new JSONArray(com.uiptv.util.ServerUtils.objectToJson(cached)));
            }
        }

        JSONObject seasonInfo = new JSONObject();
        List<String> fuzzyHints = buildFuzzyHints(seriesName, seasonInfo, response.optJSONArray("episodes"));
        JSONObject imdbFirst = ImdbMetadataService.getInstance().findBestEffortDetails(seriesName, "", fuzzyHints);
        copyIfPresent(seasonInfo, imdbFirst, "name");
        copyIfPresent(seasonInfo, imdbFirst, "cover");
        copyIfPresent(seasonInfo, imdbFirst, "plot");
        copyIfPresent(seasonInfo, imdbFirst, "cast");
        copyIfPresent(seasonInfo, imdbFirst, "director");
        copyIfPresent(seasonInfo, imdbFirst, "genre");
        copyIfPresent(seasonInfo, imdbFirst, "releaseDate");
        copyIfPresent(seasonInfo, imdbFirst, "rating");
        copyIfPresent(seasonInfo, imdbFirst, "tmdb");
        copyIfPresent(seasonInfo, imdbFirst, "imdbUrl");
        if (imdbFirst.optJSONArray("episodesMeta") != null) {
            response.put("episodesMeta", imdbFirst.optJSONArray("episodesMeta"));
        }

        if (account.getType() == AccountType.XTREME_API && !isBlank(seriesId)) {
            EpisodeList details = XtremeParser.parseEpisodes(seriesId, account);
            if (details != null) {
                SeasonInfo info = details.getSeasonInfo();
                if (info != null) {
                    JSONObject provider = new JSONObject(info.toJson());
                    // Fill missing fields only; IMDb remains preferred when available.
                    mergeMissing(seasonInfo, provider, "name");
                    mergeMissing(seasonInfo, provider, "cover");
                    mergeMissing(seasonInfo, provider, "plot");
                    mergeMissing(seasonInfo, provider, "cast");
                    mergeMissing(seasonInfo, provider, "director");
                    mergeMissing(seasonInfo, provider, "genre");
                    mergeMissing(seasonInfo, provider, "releaseDate");
                    mergeMissing(seasonInfo, provider, "rating");
                    mergeMissing(seasonInfo, provider, "tmdb");
                }

                Map<String, JSONObject> episodesMeta = indexEpisodesMeta(imdbFirst.optJSONArray("episodesMeta"));
                JSONArray episodesJson = new JSONArray();
                if (details.getEpisodes() != null) {
                    for (Episode episode : details.getEpisodes()) {
                        if (episode == null) {
                            continue;
                        }
                        Channel channel = new Channel();
                        channel.setChannelId(episode.getId());
                        channel.setName(episode.getTitle());
                        channel.setCmd(episode.getCmd());
                        channel.setExtraJson(episode.toJson());
                        channel.setSeason(episode.getSeason());
                        channel.setEpisodeNum(episode.getEpisodeNum());
                        if (episode.getInfo() != null) {
                            channel.setLogo(episode.getInfo().getMovieImage());
                            channel.setDescription(episode.getInfo().getPlot());
                            channel.setReleaseDate(episode.getInfo().getReleaseDate());
                            channel.setRating(episode.getInfo().getRating());
                            channel.setDuration(episode.getInfo().getDuration());
                            if (isBlank(channel.getSeason())) {
                                channel.setSeason(episode.getInfo().getSeason());
                            }
                        }
                        enrichEpisode(channel, episodesMeta);
                        episodesJson.put(new JSONObject(channel.toJson()));
                    }
                }
                response.put("episodes", episodesJson);
                if (episodesJson.length() > 0) {
                    SeriesEpisodeDb.get().saveAll(account, categoryId, seriesId, toChannels(episodesJson));
                }
            }
        }

        // Refine with provider fields + episode-derived hints to recover from wrong/missing portal IDs.
        fuzzyHints = buildFuzzyHints(firstNonBlank(seasonInfo.optString("name", ""), seriesName), seasonInfo, response.optJSONArray("episodes"));
        JSONObject imdbFallback = ImdbMetadataService.getInstance()
                .findBestEffortDetails(
                        firstNonBlank(seasonInfo.optString("name", ""), seriesName),
                        seasonInfo.optString("tmdb", ""),
                        fuzzyHints
                );
        mergeMissing(seasonInfo, imdbFallback, "name");
        mergeMissing(seasonInfo, imdbFallback, "cover");
        mergeMissing(seasonInfo, imdbFallback, "plot");
        mergeMissing(seasonInfo, imdbFallback, "cast");
        mergeMissing(seasonInfo, imdbFallback, "director");
        mergeMissing(seasonInfo, imdbFallback, "genre");
        mergeMissing(seasonInfo, imdbFallback, "releaseDate");
        mergeMissing(seasonInfo, imdbFallback, "rating");
        mergeMissing(seasonInfo, imdbFallback, "tmdb");
        mergeMissing(seasonInfo, imdbFallback, "imdbUrl");
        if ((response.optJSONArray("episodesMeta") == null || response.optJSONArray("episodesMeta").isEmpty())
                && imdbFallback.optJSONArray("episodesMeta") != null) {
            response.put("episodesMeta", imdbFallback.optJSONArray("episodesMeta"));
        }
        enrichEpisodesInResponse(response);
        response.put("seasonInfo", seasonInfo);
        applyNameYearFallback(seasonInfo, seriesName);

        generateJsonResponse(ex, response.toString());
    }

    private java.util.List<Channel> toChannels(JSONArray episodesJson) {
        java.util.List<Channel> channels = new java.util.ArrayList<>();
        for (int i = 0; i < episodesJson.length(); i++) {
            JSONObject obj = episodesJson.optJSONObject(i);
            if (obj == null) {
                continue;
            }
            Channel channel = Channel.fromJson(obj.toString());
            if (channel != null) {
                channels.add(channel);
            }
        }
        return channels;
    }

    private void mergeMissing(JSONObject target, JSONObject source, String key) {
        String existing = target.optString(key, "");
        if (!isBlank(existing)) {
            return;
        }
        String incoming = source.optString(key, "");
        if (!isBlank(incoming)) {
            target.put(key, incoming);
        }
    }

    private void copyIfPresent(JSONObject target, JSONObject source, String key) {
        String incoming = source.optString(key, "");
        if (!isBlank(incoming)) {
            target.put(key, incoming);
        }
    }

    private Map<String, JSONObject> indexEpisodesMeta(JSONArray episodesMeta) {
        Map<String, JSONObject> indexed = new HashMap<>();
        if (episodesMeta == null) {
            return indexed;
        }
        for (int i = 0; i < episodesMeta.length(); i++) {
            JSONObject row = episodesMeta.optJSONObject(i);
            if (row == null) continue;
            String season = safeNumeric(row.optString("season", ""));
            String episode = safeNumeric(row.optString("episodeNum", ""));
            if (!isBlank(season) && !isBlank(episode)) {
                indexed.put(season + ":" + episode, row);
            }
            String title = normalize(row.optString("title", ""));
            if (!isBlank(title)) {
                indexed.put("title:" + title, row);
            }
        }
        return indexed;
    }

    private void enrichEpisode(Channel channel, Map<String, JSONObject> episodesMeta) {
        if (episodesMeta == null || episodesMeta.isEmpty() || channel == null) {
            return;
        }
        String season = safeNumeric(channel.getSeason());
        String episode = safeNumeric(channel.getEpisodeNum());
        JSONObject meta = null;
        if (!isBlank(season) && !isBlank(episode)) {
            meta = episodesMeta.get(season + ":" + episode);
        }
        if (meta == null) {
            meta = episodesMeta.get("title:" + normalize(channel.getName()));
        }
        if (meta == null) {
            return;
        }
        if (isBlank(channel.getDescription())) {
            channel.setDescription(meta.optString("plot", ""));
        }
        if (isBlank(channel.getReleaseDate())) {
            channel.setReleaseDate(meta.optString("releaseDate", ""));
        }
        if (!isBlank(meta.optString("logo", ""))) {
            channel.setLogo(meta.optString("logo", ""));
        }
    }

    private String normalize(String value) {
        if (isBlank(value)) {
            return "";
        }
        return value.toLowerCase().replaceAll("[^a-z0-9 ]", " ").replaceAll("\\s+", " ").trim();
    }

    private String safeNumeric(String value) {
        if (isBlank(value)) return "";
        String normalized = value.replaceAll("[^0-9]", "");
        return isBlank(normalized) ? "" : normalized;
    }

    private void applyNameYearFallback(JSONObject seasonInfo, String rawSeriesName) {
        if (isBlank(rawSeriesName)) {
            return;
        }

        String trimmed = rawSeriesName.trim();
        String inferredName = trimmed.replaceAll("\\s*\\((19|20)\\d{2}\\)\\s*$", "").trim();
        String inferredYear = "";
        java.util.regex.Matcher m = java.util.regex.Pattern.compile("\\((19|20)\\d{2}\\)\\s*$").matcher(trimmed);
        if (m.find()) {
            inferredYear = m.group().replaceAll("[^0-9]", "");
        }

        if (isBlank(seasonInfo.optString("name", "")) && !isBlank(inferredName)) {
            seasonInfo.put("name", inferredName);
        }
        if (isBlank(seasonInfo.optString("releaseDate", "")) && !isBlank(inferredYear)) {
            seasonInfo.put("releaseDate", inferredYear);
        }
    }

    private void enrichEpisodesInResponse(JSONObject response) {
        if (response == null) {
            return;
        }
        JSONArray episodes = response.optJSONArray("episodes");
        JSONArray episodesMeta = response.optJSONArray("episodesMeta");
        if (episodes == null || episodes.isEmpty() || episodesMeta == null || episodesMeta.isEmpty()) {
            return;
        }
        Map<String, JSONObject> indexed = indexEpisodesMeta(episodesMeta);
        if (indexed.isEmpty()) {
            return;
        }
        for (int i = 0; i < episodes.length(); i++) {
            JSONObject row = episodes.optJSONObject(i);
            if (row == null) continue;
            Channel channel = Channel.fromJson(row.toString());
            if (channel == null) continue;
            enrichEpisode(channel, indexed);
            episodes.put(i, new JSONObject(channel.toJson()));
        }
    }

    private List<String> buildFuzzyHints(String baseTitle, JSONObject seasonInfo, JSONArray episodes) {
        List<String> hints = new ArrayList<>();
        addHint(hints, baseTitle);
        if (seasonInfo != null) {
            addHint(hints, seasonInfo.optString("name", ""));
            addHint(hints, seasonInfo.optString("plot", ""));
            addHint(hints, seasonInfo.optString("releaseDate", ""));
        }
        if (episodes != null) {
            for (int i = 0; i < Math.min(8, episodes.length()); i++) {
                JSONObject row = episodes.optJSONObject(i);
                if (row == null) continue;
                addHint(hints, row.optString("name", ""));
                addHint(hints, row.optString("releaseDate", ""));
            }
        }
        return hints;
    }

    private void addHint(List<String> hints, String value) {
        if (hints == null || isBlank(value)) {
            return;
        }
        String cleaned = value
                .replaceAll("(?i)\\b(4k|8k|uhd|fhd|hd|sd|series|movie|complete)\\b", " ")
                .replaceAll("(?i)\\bs\\d{1,2}e\\d{1,3}\\b", " ")
                .replaceAll("[\\[\\]{}()]+", " ")
                .replaceAll("\\s+", " ")
                .trim();
        if (isBlank(cleaned) || cleaned.length() < 2 || hints.contains(cleaned)) {
            return;
        }
        hints.add(cleaned);
    }

    private String firstNonBlank(String... values) {
        if (values == null) return "";
        for (String value : values) {
            if (!isBlank(value)) return value;
        }
        return "";
    }
}

File: server/src/main/kotlin/com/uiptv/server/api/json/HttpSeriesDetailsJsonServer.kt
package com.uiptv.server.api.json

import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpHandler
import com.uiptv.db.SeriesEpisodeDb
import com.uiptv.model.Account
import com.uiptv.model.Channel
import com.uiptv.service.AccountService
import com.uiptv.service.ConfigurationService
import com.uiptv.service.HandshakeService
import com.uiptv.service.ImdbMetadataService
import com.uiptv.shared.Episode
import com.uiptv.shared.EpisodeList
import com.uiptv.shared.SeasonInfo
import com.uiptv.util.AccountType
import com.uiptv.util.ServerUtils.generateJsonResponse
import com.uiptv.util.ServerUtils.getParam
import com.uiptv.util.StringUtils.isBlank
import com.uiptv.util.XtremeApiParser
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException

class HttpSeriesDetailsJsonServer : HttpHandler {
    companion object {
        private const val KEY_COVER = "cover"
        private const val KEY_DIRECTOR = "director"
        private const val KEY_EPISODES = "episodes"
        private const val KEY_EPISODES_META = "episodesMeta"
        private const val KEY_GENRE = "genre"
        private const val KEY_IMDB_URL = "imdbUrl"
        private const val KEY_RATING = "rating"
        private const val KEY_RELEASE_DATE = "releaseDate"
    }

    @Throws(IOException::class)
    override fun handle(ex: HttpExchange) {
        val account = AccountService.getInstance().getById(getParam(ex, "accountId"))
        if (account == null) {
            generateJsonResponse(ex, """{"seasonInfo":{},"$KEY_EPISODES":[]}""")
            return
        }
        if (account.isNotConnected()) {
            HandshakeService.getInstance().connect(account)
        }

        val seriesId = getParam(ex, "seriesId")
        val categoryId = getParam(ex, "categoryId")
        val seriesName = getParam(ex, "seriesName")
        val response = createBaseResponse(account, categoryId, seriesId)
        val seasonInfo = JSONObject()
        val imdbFirst = applyInitialImdbMetadata(seriesName, response, seasonInfo)
        applyProviderSeriesDetails(account, categoryId, seriesId, response, seasonInfo, imdbFirst)
        applyFallbackImdbMetadata(seriesName, response, seasonInfo)
        enrichEpisodesInResponse(response)
        response.put("seasonInfo", seasonInfo)
        applyNameYearFallback(seasonInfo, seriesName)
        generateJsonResponse(ex, response.toString())
    }

    private fun createBaseResponse(account: Account, categoryId: String?, seriesId: String?): JSONObject {
        val response = JSONObject()
        response.put("seasonInfo", JSONObject())
        response.put(KEY_EPISODES, JSONArray())
        response.put(KEY_EPISODES_META, JSONArray())
        if (isBlank(seriesId)) return response
        val resolvedCategoryId = categoryId ?: ""
        val resolvedSeriesId = seriesId ?: return response
        val cached = SeriesEpisodeDb.get().getEpisodes(account, resolvedCategoryId, resolvedSeriesId)
        if (cached.isNotEmpty() && SeriesEpisodeDb.get().isFresh(account, resolvedCategoryId, resolvedSeriesId, ConfigurationService.getInstance().getCacheExpiryMs())) {
            response.put(KEY_EPISODES, JSONArray(com.uiptv.util.ServerUtils.objectToJson(cached)))
        }
        return response
    }

    private fun applyInitialImdbMetadata(seriesName: String?, response: JSONObject, seasonInfo: JSONObject): JSONObject {
        val resolvedSeriesName = seriesName ?: ""
        val fuzzyHints = buildFuzzyHints(resolvedSeriesName, seasonInfo, response.optJSONArray(KEY_EPISODES))
        val imdbFirst = ImdbMetadataService.getInstance().findBestEffortDetails(resolvedSeriesName, "", fuzzyHints)
        copyMetadata(seasonInfo, imdbFirst)
        imdbFirst.optJSONArray(KEY_EPISODES_META)?.let { response.put(KEY_EPISODES_META, it) }
        return imdbFirst
    }

    private fun applyProviderSeriesDetails(
        account: Account,
        categoryId: String?,
        seriesId: String?,
        response: JSONObject,
        seasonInfo: JSONObject,
        imdbFirst: JSONObject
    ) {
        if (account.type != AccountType.XTREME_API || isBlank(seriesId)) {
            return
        }
        val resolvedSeriesId = seriesId ?: return
        val resolvedCategoryId = categoryId ?: ""
        val details = XtremeApiParser.parseEpisodes(resolvedSeriesId, account) ?: return
        mergeProviderSeasonInfo(seasonInfo, details.seasonInfo)
        val episodesJson = toEpisodesJson(details, indexEpisodesMeta(imdbFirst.optJSONArray(KEY_EPISODES_META)))
        response.put(KEY_EPISODES, episodesJson)
        if (episodesJson.length() > 0) {
            SeriesEpisodeDb.get().saveAll(account, resolvedCategoryId, resolvedSeriesId, toChannels(episodesJson))
        }
    }

    private fun mergeProviderSeasonInfo(seasonInfo: JSONObject, info: SeasonInfo?) {
        if (info == null) return
        mergeMetadata(seasonInfo, JSONObject(info.toJson()))
    }

    private fun toEpisodesJson(details: EpisodeList, episodesMeta: Map<String, JSONObject>): JSONArray {
        val episodesJson = JSONArray()
        details.episodes?.forEach { episode ->
            val channel = toEpisodeChannel(episode, episodesMeta)
            if (channel != null) {
                episodesJson.put(JSONObject(channel.toJson()))
            }
        }
        return episodesJson
    }

    private fun toEpisodeChannel(episode: Episode?, episodesMeta: Map<String, JSONObject>): Channel? {
        if (episode == null) return null
        val channel = Channel()
        channel.channelId = episode.id
        channel.name = episode.title
        channel.cmd = episode.cmd
        channel.extraJson = episode.toJson()
        channel.season = episode.season
        channel.episodeNum = episode.episodeNum
        val info = episode.info
        if (info != null) {
            channel.logo = info.movieImage
            channel.description = info.plot
            channel.releaseDate = info.releaseDate
            channel.rating = info.rating
            channel.duration = info.duration
            if (isBlank(channel.season)) {
                channel.season = info.season
            }
        }
        enrichEpisode(channel, episodesMeta)
        return channel
    }

    private fun applyFallbackImdbMetadata(seriesName: String?, response: JSONObject, seasonInfo: JSONObject) {
        val resolvedSeriesName = seriesName ?: ""
        val fuzzyHints = buildFuzzyHints(firstNonBlank(seasonInfo.optString("name", ""), resolvedSeriesName), seasonInfo, response.optJSONArray(KEY_EPISODES))
        val imdbFallback = ImdbMetadataService.getInstance().findBestEffortDetails(
            firstNonBlank(seasonInfo.optString("name", ""), resolvedSeriesName),
            seasonInfo.optString("tmdb", ""),
            fuzzyHints
        )
        mergeMetadata(seasonInfo, imdbFallback)
        if ((response.optJSONArray(KEY_EPISODES_META) == null || response.optJSONArray(KEY_EPISODES_META).isEmpty) &&
            imdbFallback.optJSONArray(KEY_EPISODES_META) != null) {
            response.put(KEY_EPISODES_META, imdbFallback.optJSONArray(KEY_EPISODES_META))
        }
    }

    private fun copyMetadata(target: JSONObject, source: JSONObject) {
        listOf("name", KEY_COVER, "plot", "cast", KEY_DIRECTOR, KEY_GENRE, KEY_RELEASE_DATE, KEY_RATING, "tmdb", KEY_IMDB_URL)
            .forEach { copyIfPresent(target, source, it) }
    }

    private fun mergeMetadata(target: JSONObject, source: JSONObject) {
        listOf("name", KEY_COVER, "plot", "cast", KEY_DIRECTOR, KEY_GENRE, KEY_RELEASE_DATE, KEY_RATING, "tmdb", KEY_IMDB_URL)
            .forEach { mergeMissing(target, source, it) }
    }

    private fun toChannels(episodesJson: JSONArray): List<Channel> {
        val channels = ArrayList<Channel>()
        for (i in 0 until episodesJson.length()) {
            val obj = episodesJson.optJSONObject(i) ?: continue
            Channel.fromJson(obj.toString())?.let(channels::add)
        }
        return channels
    }

    private fun mergeMissing(target: JSONObject, source: JSONObject, key: String) {
        val existing = target.optString(key, "")
        if (!isBlank(existing)) return
        val incoming = source.optString(key, "")
        if (!isBlank(incoming)) target.put(key, incoming)
    }

    private fun copyIfPresent(target: JSONObject, source: JSONObject, key: String) {
        val incoming = source.optString(key, "")
        if (!isBlank(incoming)) target.put(key, incoming)
    }

    private fun indexEpisodesMeta(episodesMeta: JSONArray?): Map<String, JSONObject> {
        val indexed = HashMap<String, JSONObject>()
        if (episodesMeta == null) return indexed
        for (i in 0 until episodesMeta.length()) {
            val row = episodesMeta.optJSONObject(i) ?: continue
            val season = safeNumeric(row.optString("season", ""))
            val episode = safeNumeric(row.optString("episodeNum", ""))
            if (!isBlank(season) && !isBlank(episode)) indexed["$season:$episode"] = row
            val title = normalize(row.optString("title", ""))
            if (!isBlank(title)) indexed["title:$title"] = row
        }
        return indexed
    }

    private fun enrichEpisode(channel: Channel?, episodesMeta: Map<String, JSONObject>?) {
        if (episodesMeta.isNullOrEmpty() || channel == null) return
        val season = safeNumeric(channel.season)
        val episode = safeNumeric(channel.episodeNum)
        var meta: JSONObject? = null
        if (!isBlank(season) && !isBlank(episode)) {
            meta = episodesMeta["$season:$episode"]
        }
        if (meta == null) {
            meta = episodesMeta["title:${normalize(channel.name)}"]
        }
        if (meta == null) return
        if (isBlank(channel.description)) channel.description = meta.optString("plot", "")
        if (isBlank(channel.releaseDate)) channel.releaseDate = meta.optString(KEY_RELEASE_DATE, "")
        if (!isBlank(meta.optString("logo", ""))) channel.logo = meta.optString("logo", "")
    }

    private fun normalize(value: String?): String =
        if (isBlank(value)) "" else value!!.lowercase().replace(Regex("[^a-z0-9 ]"), " ").replace(Regex("\\s+"), " ").trim()

    private fun safeNumeric(value: String?): String {
        if (isBlank(value)) return ""
        val normalized = value!!.replace(Regex("\\D"), "")
        return if (isBlank(normalized)) "" else normalized
    }

    private fun applyNameYearFallback(seasonInfo: JSONObject, rawSeriesName: String?) {
        if (isBlank(rawSeriesName)) return
        val trimmed = rawSeriesName?.trim() ?: return
        val inferredName = trimmed.replace(Regex("\\s*\\((19|20)\\d{2}\\)\\s*$"), "").trim()
        var inferredYear = ""
        val matcher = Regex("\\((19|20)\\d{2}\\)\\s*$").find(trimmed)
        if (matcher != null) {
            inferredYear = matcher.value.replace(Regex("\\D"), "")
        }
        if (isBlank(seasonInfo.optString("name", "")) && !isBlank(inferredName)) seasonInfo.put("name", inferredName)
        if (isBlank(seasonInfo.optString(KEY_RELEASE_DATE, "")) && !isBlank(inferredYear)) seasonInfo.put(KEY_RELEASE_DATE, inferredYear)
    }

    private fun enrichEpisodesInResponse(response: JSONObject?) {
        if (response == null) return
        val episodes = response.optJSONArray(KEY_EPISODES)
        val episodesMeta = response.optJSONArray(KEY_EPISODES_META)
        if (episodes == null || episodes.isEmpty || episodesMeta == null || episodesMeta.isEmpty) return
        val indexed = indexEpisodesMeta(episodesMeta)
        if (indexed.isEmpty()) return
        for (i in 0 until episodes.length()) {
            val row = episodes.optJSONObject(i) ?: continue
            val channel = Channel.fromJson(row.toString()) ?: continue
            enrichEpisode(channel, indexed)
            episodes.put(i, JSONObject(channel.toJson()))
        }
    }

    private fun buildFuzzyHints(baseTitle: String?, seasonInfo: JSONObject?, episodes: JSONArray?): List<String> {
        val hints = ArrayList<String>()
        addHint(hints, baseTitle)
        if (seasonInfo != null) {
            addHint(hints, seasonInfo.optString("name", ""))
            addHint(hints, seasonInfo.optString("plot", ""))
            addHint(hints, seasonInfo.optString(KEY_RELEASE_DATE, ""))
        }
        if (episodes != null) {
            for (i in 0 until minOf(8, episodes.length())) {
                val row = episodes.optJSONObject(i) ?: continue
                addHint(hints, row.optString("name", ""))
                addHint(hints, row.optString(KEY_RELEASE_DATE, ""))
            }
        }
        return hints
    }

    private fun addHint(hints: MutableList<String>, value: String?) {
        if (isBlank(value)) return
        val cleaned = value!!
            .replace(Regex("(?i)\\b(4k|8k|uhd|fhd|hd|sd|series|movie|complete)\\b"), " ")
            .replace(Regex("(?i)\\bs\\d{1,2}e\\d{1,3}\\b"), " ")
            .replace(Regex("[\\[\\]{}()]+"), " ")
            .replace(Regex("\\s+"), " ")
            .trim()
        if (isBlank(cleaned) || cleaned.length < 2 || hints.contains(cleaned)) return
        hints += cleaned
    }

    private fun firstNonBlank(vararg values: String?): String = values.firstOrNull { !isBlank(it) } ?: ""
}
